mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/PostgresMessageBlobReferenceSource.java

```diff
@@ -0,0 +1,42 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mailbox.postgres.mail;
+
+import javax.inject.Inject;
+
+import org.apache.james.blob.api.BlobId;
+import org.apache.james.blob.api.BlobReferenceSource;
+import org.apache.james.mailbox.postgres.mail.dao.PostgresMessageDAO;
+
+import reactor.core.publisher.Flux;
+
+public class PostgresMessageBlobReferenceSource implements BlobReferenceSource {
+    private PostgresMessageDAO postgresMessageDAO;
+
+    @Inject
+    public PostgresMessageBlobReferenceSource(PostgresMessageDAO postgresMessageDAO) {
+        this.postgresMessageDAO = postgresMessageDAO;
+    }
+
+    @Override
+    public Flux<BlobId> listReferencedBlobs() {
+        return postgresMessageDAO.listBlobs();
```

Comment:
For byPassRLS
we can:
- list all domain
- for each domain, list all blobId
- concat blobId

WDYT?
Alternative idea:
- 1. Create a mid-table for storage blobId without domain columns (for bypass RLS)
- 2. Create a "super" Connection (by super User?) for bypass RLS

Reply:
No Hung approach seems more straightforward to me: obtain a root session for the dedup purpose...

Reply:
> have a complexity O(n*n) as for each domain

it shouldn't even if it is a task?

To create a "root session", the basic way is configuration properties support declaring an RLS user, super User, then we can create a root session based on the superuser. 
Not sure can have another easier way?

Reply:
What's wrong with having a super user creating a root session? as long as it's just used internally I think we good?

Reply:
> What's wrong with having a super user creating a root session? as long as it's just used internally I think we good?

No Wrong, it's just more code, I'm ok with that, it is also my think
> Alternative idea:
> Create a "super" Connection (by super User?) for bypass RLS

Reply:
Ah yes fair :) The first idea is more requests to the db though to list all blobs. I would not be surprised as well that we find more cases later where a root session would be useful and save us from more sql requests too, so I don't think it's bad to invest a bit of time on it now :)

Reply:
I have not done anything yet in this ticket to solve this issue. I have just discussed with Tung about this to find solution. Not sure how you @chibenwa obtain root session. For me, the only way to get it is using root user or an user that have privilege of bypassing rls to obtain a session that can bypass RLS. And to achieve that, we have to create two postgresqlConnectionFactory(s):

```
new PostgresqlConnectionFactory(PostgresqlConnectionConfiguration.builder()
            .host(postgresConfiguration.getUri().getHost())
            .port(postgresConfiguration.getUri().getPort())
            .username(postgresConfiguration.getCredential().getUsername())
            .password(postgresConfiguration.getCredential().getPassword())
            .database(postgresConfiguration.getDatabaseName())
            .schema(postgresConfiguration.getDatabaseSchema())
            .build()); 
```
one PostgresqlConnectionFactory has been used so far. A new one would use root user or an user that have privilege of bypassing rls.

WDYT?

Reply:
fixed